src/source/Response.c

```diff
@@ -625,26 +648,42 @@ SIZE_T postReadCallback(PCHAR pBuffer, SIZE_T size, SIZE_T numItems, PVOID custo
     pCurlApiCallbacks = pCurlRequest->pCurlApiCallbacks;
     uploadHandle = pCurlResponse->pCurlRequest->uploadHandle;
 
-    if (pCurlResponse->paused) {
-        bytesWritten = CURL_READFUNC_PAUSE;
-        CHK(FALSE, retStatus);
-    }
-
     if (pCurlResponse->endOfStream || ATOMIC_LOAD_BOOL(&pCurlRequest->requestInfo.terminating)) {
-        DLOGI("Closing connection for upload stream handle: %" PRIu64, uploadHandle);
+        DLOGI("Closing connection for upload stream handle: %"
+        PRIu64, uploadHandle);
         CHK(FALSE, retStatus);
     }
 
-    retStatus =
-        getKinesisVideoStreamData(pCurlResponse->pCurlRequest->streamHandle, uploadHandle, (PBYTE) pBuffer, (UINT32) bufferSize, &retrievedSize);
+    retStatus = getKinesisVideoStreamData(pCurlResponse->pCurlRequest->streamHandle, uploadHandle, (PBYTE) pBuffer,
+                                          (UINT32) bufferSize, &retrievedSize);
 
     if (pCurlApiCallbacks->curlReadCallbackHookFn != NULL) {
         retStatus =
-            pCurlApiCallbacks->curlReadCallbackHookFn(pCurlResponse, uploadHandle, (PBYTE) pBuffer, (UINT32) bufferSize, &retrievedSize, retStatus);
+                pCurlApiCallbacks->curlReadCallbackHookFn(pCurlResponse, uploadHandle, (PBYTE) pBuffer,
+                                                          (UINT32) bufferSize, &retrievedSize, retStatus);
     }
 
     bytesWritten = (SIZE_T) retrievedSize;
 
+    // If conditions are met under which we would have previously called curl pause -- instead wait on CVAR
+    // SIGNAL is in notifyDataAvailable
+    if (bytesWritten == 0 && (retStatus == STATUS_SUCCESS || retStatus == STATUS_NO_MORE_DATA_AVAILABLE || retStatus == STATUS_AWAITING_PERSISTED_ACK)) {
+        pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.lockMutexFn(pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.customData, pCurlResponse->dataAvailableLock);
+        CVAR_WAIT(pCurlResponse->dataAvailableCvar, pCurlResponse->dataAvailableLock, TIMEOUT_TEARDOWN_CURL_NO_DATA);
+        pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.unlockMutexFn(pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.customData, pCurlResponse->dataAvailableLock);
+
+        retStatus = getKinesisVideoStreamData(pCurlResponse->pCurlRequest->streamHandle, uploadHandle, (PBYTE) pBuffer,

```

Comment:
Perhaps, the approach should be something like this

mutex lock
while (bytesWritten == 0) {
    awaitStatus = CVAR_WAIT();
    // Deal with a timeout
    if (awaitStatus == TIMEOUT) {
        retStatus = terminate/abort; ???
    }

    // Proceed with call to get data
    retStatus = getStreamData();
    Proceed with handling of the bytes - switch case with possible abort on bad retStatus
}
mutex unlock